### PR TITLE
feat: ING Bank Australia

### PR DESCRIPTION
**Pull request by Scrub000**
## What does this PR do?

This PR introduces the ING Bank Australia piece, which allows a user to download their bank statements in multiple formats, such as CSV or QIF.

These statements can then be added to e.g. Actual Budget ([piece](https://github.com/activepieces/activepieces/pull/5093)) or Firefly III.

Screenshot_20240723_081845

Credit to https://github.com/adamroyle/ing-au-login, for the code to log into ING.

## Dependencies

In order for this PR to work in ActivePieces, the following dependencies will need to be installed:

* NPM: `npm install ing-au-login puppeteer`
* APT, in the Dockerfile: `sudo apt install libnss3 libatk1.0-0 libatk-bridge2.0-0 libcups2 libgbm1 libasound2 libpangocairo-1.0-0 libxss1 libgtk-3-0` (https://gist.github.com/tavinus/7effd4b3dac87cb4366e3767679a192f) - For the headless browser.
* ING Bank Australia logo will need to be re-uploaded to a permanent location.

### CODE REVIEW

File: packages/pieces/community/ing-australia/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+          return {
+              valid: true,
+          }
+      }
+      return {
+          valid: false,
+          error: 'Invalid Api Key'
+      }
+  }
+});
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/src/index.ts:27]: TODO: Do actual validation of card number and PIN code.

File: packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      args: ['--no-sandbox', '--disable-setuid-sandbox'],
+    });
+    const page = await browser.newPage();
+    const authToken = await login(
+      page,
+      context.auth.username,
+      context.auth.password
+    );
+    await page.close();
+    await browser.close();
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts:39]: TODO: Move this into its own function. Need to make sending API requests more streamlined. Also need to cache this for a bit. Need to ensure we close the browser process, always. Got to catch exceptions.

File: packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    // await browser.close();
+    // return await httpClient.sendRequest<never>(request);
+
+    const request: HttpRequest = {
+      method: HttpMethod.POST,
+      url: 'https://www.ing.com.au/api/ExportTransactions/Service/ExportTransactionsService.svc/json/ExportTransactions/ExportTransactions',
+      body: new URLSearchParams({
+        'X-AuthToken': authToken,
+        'AccountNumber': context.propsValue['account_number'],
+        'Format': 'csv',
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts:61]: TODO: Support a drop-down of formats.

File: packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+export const downloadTransactions = createAction({
+  // auth: check https://www.activepieces.com/docs/developers/piece-reference/authentication,
+  name: 'downloadTransactions',
+  auth: ingAustraliaAuth,
+  displayName: 'Download Transactions',
+  description:
+    'Download bank transactions in multiple formats, such as CSV or OFX.',
+  props: {
+    account_number: Property.ShortText({
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts:21]: Validate props

File: packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    //   },
+    // };
+    // await browser.close();
+    // return await httpClient.sendRequest<never>(request);
+
+    const request: HttpRequest = {
+      method: HttpMethod.POST,
+      url: 'https://www.ing.com.au/api/ExportTransactions/Service/ExportTransactionsService.svc/json/ExportTransactions/ExportTransactions',
+      body: new URLSearchParams({
+        'X-AuthToken': authToken,
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/src/lib/actions/download-transactions.ts:59]: TODO: Make this query work...

File: packages/pieces/community/ing-australia/package.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+  "name": "@activepieces/piece-ing-australia",
+  "version": "0.0.1",
+  "dependencies": {
+    "ing-au-login": "2.0.7",
+    "puppeteer": "22.13.1",
```
Comment (Scrub000) [inline packages/pieces/community/ing-australia/package.json:6]: Does ActivePieces automatically install these dependencies? In the development environment, I need to manually install them.

File: .devcontainer/Dockerfile [hunk 7]
```diff
@@ -3,8 +3,17 @@ ARG VARIANT=18-bullseye
...
+    && apt-get -y install --no-install-recommends \
+    libnss3 \
+    libatk1.0-0 \
+    libatk-bridge2.0-0 \
+    libcups2 \
+    libgbm1 \
+    libasound2 \
+    libpangocairo-1.0-0 \
+    libxss1 \
+    libgtk-3-0
```
Comment (Scrub000) [inline .devcontainer/Dockerfile:16]: I've added these dependencies to both the development and production Dockerfiles, as they're needed to support opening a headless browser.

The development Dockerfile appears to work with these dependencies, however I note the image is different in production so I'm not sure if it'll work? `FROM node:18.19-bullseye-slim AS base` vs `ARG VARIANT=18-bullseye
FROM mcr.microsoft.com/vscode/devcontainers/javascript-node:0-${VARIANT}`

**Comment by Scrub000** [comment]
Closing; Security concerns around unsandboxed browser.

**Comment by Scrub000** [comment]
Reopening; New workaround will be to require that the user deploy browserless on their machine, and to then provide the endpoint in the authentication step.

**Comment by abuaboud** [comment]
Hi @Scrub000 

Thank you for providing the update, I would suggest start with publishing to the npm directly as community piece.

I think this can be reopened later once we have official support for playwright and puppeteer for all execution mode types. 

What do you think?

**Pull request closed without merging**